src/base/ClpMap.h

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright (C) 1996-2020 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID__SRC_BASE_CLPMAP_H
+#define SQUID__SRC_BASE_CLPMAP_H
+
+#include "SquidTime.h"
+
+#include <list>
+#include <map>
+
+template<class EntryValue>
+size_t
+DefaultMemoryUsage(const EntryValue *e)
+{
+    return sizeof(*e);
+}
+
+/**
+ * A Map for caching data by Capacity, Lifetime, and Priority (CLP)
+ * Unlike other Map containers data is;
+ * - added only if it fits within a predetermined memory limit (Capacity),
+ * - gets expired based on TTL (Lifetime), and a fading Priority Queue.
+ */
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *) = DefaultMemoryUsage>
+class ClpMap
+{
+public:
+    class Entry
+    {
+    public:
+        Entry(const Key &aKey, EntryValue *t): key(aKey), value(t), date(squid_curtime) {}
+        ~Entry() {delete value;}
+    private:
+        Entry(Entry &);
+        Entry & operator = (Entry &);
+    public:
+        Key key; ///< the key of entry
+        EntryValue *value = nullptr; ///< A pointer to the stored value
+        time_t date = 0; ///< The date the entry created
+    };
+    typedef std::list<Entry *> Queue;
+    typedef typename std::list<Entry *>::iterator QueueIterator;
+
+    /// key:queue_item mapping for fast lookups by key
+    typedef std::map<Key, QueueIterator> Map;
```

Comment:
Why do we need to order QueueIterators by their keys here? Maintaining that order usually costs a lot of extra CPU cycles (`O(log(n))` instead of `O(1)` of std::unordered_map). What makes this use case special?

Reply:
Just that this is the way LruMap did things. This container has the same properties that one had. The adjustments made by this PR only affect the overall maximum N - by reducing it in most installations, though for some increasing it.

Reply:
Understood, thank you. It is my understanding that this PR replaces LruMap rather than enhances it. If my understanding is correct, we should evaluate ClpMap design. This particular design aspect looks questionable to me (for the reasons I have stated in the change request). If you share my concern, please address it. Otherwise, let's discuss why my concern is invalid.

Reply:
The rename does imply replacement, but the implemented logic for now is still LRU in the queue mechanism.

The rename is to remove the implication that the queue type has to be an LRU. Resolving that sticking point in our previous disagreements. The LRU portion can be replaced in future with any other queue ordering and the type remains ClpMap.

Reply:
Here is a specific suggestion that would move us towards addressing this old change request:

```suggestion
    /// key:queue_item mapping for fast lookups by key
    typedef std::unordered_map<Key, QueueIterator, ..., ..., PoolingAllocator<MapItem> > Map;
```

... but the std::unordered_map template will need a hash and equality functors that I cannot just type in from memory here. This is the primary reason why I hesitated providing an exact code line after you confirmed that we do not need the O(log(n)) search overhead here.

BTW, the MapPair type declared below should probably be declared above. I recommend calling it MapItem because the "pair" is an unimportant implementation detail that may change.